### PR TITLE
Add webpack 4 plugin api support

### PR DESCRIPTION
**Pull request by log2-hwan**
... and remove deprecation warnings for webpack 4 users

### CODE REVIEW

**Comment by kfeinUI** [comment]
Looking forward to trying it out.

**Comment by slaby93** [comment]
Any progress?

File: lib/HappyPlugin.js [hunk 1]
```diff
@@ -66,9 +66,10 @@ function HappyPlugin(userConfig) {
 }
 
 HappyPlugin.prototype.apply = function(compiler) {
-  var that, engageWatchMode;
+  var that, webpack4PluginSupport, engageWatchMode, watchRunCallback, doneCallback;
 
   that = this;
+  webpack4PluginSupport = 'hooks' in compiler;
```
Comment (amireh) [inline lib/HappyPlugin.js]: I'd rather have an abstraction that decides whether to call the `plugin` function directly or to use the `hooks` API instead.

A sample implementation would be:

```javascript
function tapInto(target, hook, f) {
  if (webpack4) {
    const webpack4Hook = resolveHook(hook) // e.g. watch-run -> watchRun
    target.hooks[webpack4Hook].tap('HappyPlugin', f)
  } else {
    target.plugin(hook, f)
}

tapInto(compiler, 'watch-run', ...)
tapInto(compiler, 'compilation', compilation => {
  tapInto(compilation, 'failed-module', ...)
})
```

That would help keep the code remain readable.

**Comment by log2-hwan** [comment]
@amireh I updated it based on your suggestion. Could you review it again?

**Comment by amireh** [comment]
Thank you! We could probably add a different hook for async taps but I'll do that myself before the release.

**Pull request merged**